scripts/py_matter_yamltests/matter_yamltests/constraints.py

```diff
@@ -582,7 +582,11 @@ def __init__(self, context, is_upper_case):
         self._is_upper_case = is_upper_case
 
     def check_response(self, value, value_type_name) -> bool:
-        return value.isupper() == self._is_upper_case
+        # isupper() on the whole string will return false for digit-only
+        # strings, which is not what we want here.  We want to treat digits as
+        # uppercase for purposes of isUpperCase.
+        isUpper = all(c.isdigit() or c.isupper() for c in value)
```

Comment:
for symmetry, why don't we have 

```
hasLower == any(c.islower() for c in value)
return hasLower != self.is_upper_case

Reply:
Because I tried to implement exatly what the C++ constraint checks do right now.  It actually looks like I did not quite succeed; for empty string C++ will test true for isUpperCase, but this code tests false.

Realistically, we need to decide how to handle the following for isUpperCase/isLowerCase:

1. Empty strings.
2. Strings with only digits.
3. Strings with non-digit-and-non-letter characters (e.g. punctuation).

@vivien-apple

Reply:
I would say that all characters that does not have a lowercase/uppercase equivalent should not affect the output of the method. So, all the 3 mentioned cases should be true for both in our case.

Alternatively, if we want to be stricter and check if the string contains *only* uppercase characters that has a lowercase version, we can add a new constraint (e.g isUpperCaseCharactersOnly).

Reply:
Did that, added testcases.